### PR TITLE
Consolidate use of Foam URI

### PR DESCRIPTION
**Pull request by riccardoferretti**
With this PR all URIs that Foam will work with in its internals will be Foam URI, as opposed to also VS Code Uri.

A few related issues: #813 #814 #816 
VS Code related issues: https://github.com/microsoft/vscode/issues/43959 https://github.com/microsoft/vscode/issues/116298 https://github.com/microsoft/vscode/issues/42159 https://github.com/microsoft/vscode/issues/70004 https://github.com/microsoft/vscode/issues/67523 https://github.com/microsoft/vscode/issues/12448

### The problem

Basically VS Code is not consistent in the way they handle the casing of the drive letter for Windows, and this is causing issues in Foam.

Originally we wanted Foam `URI`, and VS Code `Uri` to be interchangeable (or at least for Foam to be able to consume VS Code `Uri`), but these internal inconsistencies (with no plan to be fixed) have made working in Windows especially error prone.

### The solution

With this PR:
- VS Code `Uri`s are always converted to Foam `URI` before being used by our API
- Foam `URI` always normalizes Windows paths (uppercase drive letter, following the VS Code recommendation, see https://github.com/microsoft/vscode/issues/43959)
- Updated tests

### Considerations

- I actually don't mind what the code looks like, and the increased clarity and consistency are well worth the work
- 

## Challenges

There are still some challenges with this approach.

Originally we wanted `URI => Uri => URI` and `Uri => URI => Uri` to be idempotent, obviously that won't be the case moving forward (but as we will always use `URI` internally, we only have to worry about this on the edges - whereas the current approach makes everything brittle)

Going the route of formally splitting `foam.URI` and `vscode.Uri` has an implementation challenge because for this to make sense typescript has to enforce it, but typescript uses duck typing and I haven't found a clean way to force the conversion. The only options I can think of atm are: renaming one of the fields, adding extra ones, or artificially being more restrictive on the types of existing ones.
Related to this point: microsoft/TypeScript#42698 which means we could use classes in the future once that issue is resolved.

This PR of course would require thorough testing in Windows.

## What reviewers should focus on

- considerations and challenges from the PR
- at a code level, whether the approach makes sense
- test coverage / edge cases
- if you have a windows machine, please play with this 🙏

### CODE REVIEW

**Review by movermeyer — APPROVED** [review]
Reasoning seems sensible. Code seems good. 👍

File: packages/foam-vscode/src/features/create-from-template-variables.spec.ts [hunk 1]
```diff
@@ -129,7 +130,7 @@ describe('resolveFoamVariables', () => {
   });
 
   test('Resolves FOAM_DATE_* properties with given date', async () => {
-    const targetDate = new Date(2021, 8, 12, 1, 2, 3);
+    const targetDate = new Date(2021, 9, 12, 1, 2, 3);
```
[the remark this answers is not in the document]
Comment (riccardoferretti) [reply packages/foam-vscode/src/features/create-from-template-variables.spec.ts:133]: ooops, that is completely incidental, my bad - but not a big deal so let's keep it

**Comment by AndreiD049** [comment]
Hey, i played a little with this branch locally, and it seemed fine to me.
2 things i noticed that were kind of strnage:
1. Hovering - if i have more links to a note, it appears duplicated when hovering.
image
While my todo looks like this: 
image

2. Also Hovering. Let's say i have the same situation like in case nr. 1, but `[[new-placeholder]]` is not yet created. Then i `Ctrl+click` on it to create it, make some changes in `inbox` and `todo`, and then when i hover over `[[new-placeholder]]` in `inbox`, i see it being double-doubled 🤔
image

This is not particularly related to this PR, i can replicate the same behavior on `master`, just my 2c.
Besides that, looks good 👍
